timeout_test.go

```diff
@@ -0,0 +1,177 @@
+package graphql_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	graphql "github.com/graph-gophers/graphql-go"
+	qerrors "github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/gqltesting"
+)
+
+type TimeoutTest struct {
+}
+
+func (r *TimeoutTest) Message(args struct{ Timeout int32 }) *string {
+	time.Sleep(time.Duration(args.Timeout) * time.Millisecond)
+	s := "Success!"
+	return &s
+}
+
+func (r *TimeoutTest) MetaMessage() MetaMessage {
+	m := MetaMessage{}
+	return m
+}
+
+type MetaMessage struct {
+}
+
+func (m MetaMessage) Msg(args struct{ Timeout int32 }) string {
+	time.Sleep(time.Duration(args.Timeout) * time.Millisecond)
+	s := "MetaSuccess!"
+	return s
+}
+
+func TestPartialSuccessTimeout(t *testing.T) {
+	ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(20*time.Millisecond)) // This test now depends on the simplest resolver returning within 10 milliseconds
+	gqltesting.RunTest(t,
+		&gqltesting.Test{ // test that one feature will sucessfully return
+			Schema: graphql.MustParseSchema(`
+					schema {
+						query: Query
+					}
+
+					type Query {
+						Message(Timeout: Int!): String
+					}
+				`, &TimeoutTest{}),
+			Query: `
+					{
+					m1: Message(Timeout: 1)
+					m2: Message(Timeout: 2000)
+					}
+				`,
+			ExpectedResult: ` { "m1": "Success!", "m2": null }`,
+			ExpectedErrors: []*qerrors.QueryError{
+
+				{
+					Message: "context deadline exceeded",
+					Path:    []interface{}{"m2"},
+				},
+			},
+			Context: ctx,
+		})
+}
+
+func TestPartialSuccessCancel(t *testing.T) {
+
+	ctx2, cancelFunc := context.WithDeadline(context.Background(), time.Now().Add(50*time.Millisecond))
+	go func() {
+		time.Sleep(20 * time.Millisecond)
```

Comment:
I don't like seeing `time.Sleep` in tests. You can mock the behaviour of a timeout. Or you can cancel the context which should result in similar behaviour, non?

Reply:
That's a fair concern, I can switch to only doing cancels. They do give different error messages, but there is no switching based off of the actual error text.